File: facade/src/main/java/org/tdf/sunflower/state/AbstractBuiltIn.java
package org.tdf.sunflower.state;

import lombok.Getter;
import org.tdf.common.util.FastByteComparisons;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.facade.RepositoryReader;
import org.tdf.sunflower.facade.RepositoryService;
import org.tdf.sunflower.types.Header;
import org.tdf.sunflower.vm.Backend;
import org.tdf.sunflower.vm.CallContext;
import org.tdf.sunflower.vm.CallData;
import org.tdf.sunflower.vm.abi.Abi;

import java.util.List;
import java.util.Objects;

public abstract class AbstractBuiltIn implements BuiltinContract {
    protected StateTrie<HexBytes, Account> accounts;
    protected RepositoryService repo;

    @Getter
    protected HexBytes address;

    protected AbstractBuiltIn(
        HexBytes address,
        StateTrie<HexBytes, Account> accounts,
        RepositoryService repo
    ) {
        this.address = address;
        this.accounts = accounts;
        this.repo = repo;
    }

    protected byte[] getSelector(HexBytes data) {
        return data.slice(0, 4).getBytes();
    }

    @Override
    public byte[] call(RepositoryReader rd, Backend backend, CallContext ctx, CallData callData) {
        Abi.Function func = getFunction(callData.getData());
        List<?> inputs = func.decode(callData.getData().getBytes());
        List<?> results = call(rd, backend, ctx, callData, func.name, inputs.toArray());
        return Abi.Entry.Param.encodeList(func.outputs, results.toArray());
    }

    protected Abi.Function getFunction(HexBytes data) {
        byte[] selector = getSelector(data);
        Abi abi = getAbi();
        Abi.Function func = Objects.requireNonNull(
            abi.findFunction(x -> FastByteComparisons.equal(x.encodeSignature(), selector))
        );
        return func;
    }

    protected Abi.Function getFunction(String method) {
        Abi abi = getAbi();
        Abi.Function func = Objects.requireNonNull(
            abi.findFunction(x -> x.name.equals(method))
        );
        return func;
    }

    @Override
    public List<?> view(RepositoryReader rd, HexBytes blockHash, String method, Object... args) {
        Header parent = rd.getHeaderByHash(blockHash);
        Abi.Function func = getFunction(method);
        byte[] encoded = func.encode(args);
        CallData callData = Utils.INSTANCE.createCallData(encoded);
        return call(rd, accounts.createBackend(parent, null, true, parent.getStateRoot()), CallContext.empty(), callData, method, args);
    }
}

File: facade/src/main/java/org/tdf/sunflower/state/AbstractBuiltin.kt
package org.tdf.sunflower.state

import org.tdf.common.util.FastByteComparisons
import org.tdf.common.util.HexBytes
import org.tdf.common.util.hex
import org.tdf.common.util.selector
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.vm.Backend
import org.tdf.sunflower.vm.CallContext
import org.tdf.sunflower.vm.CallData
import org.tdf.sunflower.vm.abi.Abi


abstract class AbstractBuiltin protected constructor(
    override val address: HexBytes,
) : Builtin {

    private fun getSelector(data: HexBytes): ByteArray {
        return data.bytes.selector()
    }

    override fun call(rd: RepositoryReader, backend: Backend, ctx: CallContext, callData: CallData): ByteArray {
        val func = getFunction(callData.data)
        val inputs: Array<Any> = func.decode(callData.data.bytes).map { it as Any }.toTypedArray()
        val results = call(rd, backend, ctx, callData, func.name, *inputs)
        return Abi.Entry.Param.encodeList(func.outputs, *results.toTypedArray())
    }

    private fun getFunction(data: HexBytes): Abi.Function {
        val selector = getSelector(data)
        return abi.findFunction { FastByteComparisons.equal(it.encodeSignature(), selector) }!!
    }

    private fun getFunction(method: String): Abi.Function {
        return abi.findFunction { it.name == method }!!
    }

    override fun view(rd: RepositoryReader, backend: Backend, method: String, vararg args: Any): List<*> {
        val func = getFunction(method)
        val encoded = func.encode(*args)
        val callData = CallData(data = encoded.hex())
        return call(
            rd,
            backend,
            CallContext(),
            callData,
            method,
            *args
        )
    }
}